### PR TITLE
[16.0][MIG] purchase_delivery_split_date: Migration to 16.0

### PR DESCRIPTION
**Pull request by FernandoRomera**

### CODE REVIEW

**Comment by FernandoRomera** [comment]
#1585

**Review by AntoniRomera — APPROVED** [review]

**Review by MRomeera — APPROVED** [review]
Ok

**Comment by BT-anieto** [comment]
Hello @FernandoRomera! Thank you for the migration.

I see two points:

1. Code coverage (you may need to create a small tests for not covered lines).
2. Squash administrative commits: (You can see this [here](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-16.0)):

`Squash administrative commits (if any) with the previous commit for reducing commit noise. They are named as "[UPD] README.rst", "[UPD] Update $MODULE.pot", "Update translation files" and similar names, and comes from OCA-git-bot, oca-travis or oca-transbot. IMPORTANT: Don't squash legit translation commits, authored by their translators, with the message "Translated using Weblate (...)".`

**Comment by FernandoRomera** [comment]
@BT-anieto 
Squash administrative commits: Done

**Comment by FernandoRomera** [comment]
@dreispt 
Can you merge this PR, please?

**Comment by BT-anieto** [comment]
@FernandoRomera Hello Fernando :) You need to increase the code coverage to merge this PR. I suggest you to create a very small test which goes through some of the not covered lines. As an example, you could create a purchase order with an order line with a form, so that the onchanges are triggered, or call them manually on the creation of the order line.

**Comment by rousseldenis** [comment]
@FernandoRomera Some things seem weird to me.

I will do a PR on your branch soon.

**Comment by rousseldenis** [comment]
@gurneyalex Do you remember why you did this and why in this module (it seems functionally out of scope)?

https://github.com/OCA/purchase-workflow/commit/b3e20e85da03b0a5ca04303943130add135019dd

**Comment by rousseldenis** [comment]
@FernandoRomera I did this : https://github.com/MallorcaSoft/purchase-workflow/pull/1

- Changed test setup to classmethod
- Fixed unused onchange to replaced compute method for date_planned field (see my previous comment)
- Split one model per file

**Comment by rousseldenis** [comment]
I think I will also modify the code on purchase order as there are some operations that can be done by batch

**Comment by rousseldenis** [comment]
> I think I will also modify the code on purchase order as there are some operations that can be done by batch

Done in PR mentionned above. I think that one is ready

**Comment by FernandoRomera** [comment]
@rousseldenis 
Done.

**Review by rousseldenis — APPROVED** [review]

**Comment by rousseldenis** [comment]
And functionally works as expected

File: purchase_delivery_split_date/models/purchase_order_line.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            moves += super(PurchaseOrderLine, po_lines)._create_stock_moves(picking)
+        return moves
+
+    def write(self, values):
+        res = super().write(values)
+        if "date_planned" in values:
+            self.mapped("order_id")._check_split_pickings()
+        return res
+
+    def create(self, values):
```
Comment (bizzappdev) [inline purchase_delivery_split_date/models/purchase_order_line.py:85]: @api.model_create_multi

File: purchase_delivery_split_date/models/purchase_order_line.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        line = super().create(values)
+        if line.order_id.state == "purchase":
+            line.order_id._check_split_pickings()
+        return line
```
Comment (lmignon) [inline purchase_delivery_split_date/models/purchase_order_line.py:89]:
```suggestion
    @api.model_create_multi
    def create(self, vals_list):
        lines = super().create(vals_list)
        lines.order_id.filtered(lambda l: l.state == "purchase")._check_split_pickings()
        return lines
```

**Comment by gurneyalex** [comment]
> @gurneyalex Do you remember why you did this and why in this module (it seems functionally out of scope)?
> 
> [b3e20e8](https://github.com/OCA/purchase-workflow/commit/b3e20e85da03b0a5ca04303943130add135019dd)

This does not seem to be out of scope to me. 

Change 1:

consider this scenario: 
* User creates a PO, sets different dates on the lines and confirms -> you get different receptions based on the date lines. 
* User adds a line in the PO, saves -> I expect the delivery of the stock move to be done according to the date of the line and this was not done prior to the fix unless the line's planned date had been set / changed

Change 2: 

scenario:
* User creates a PO, sets dates on the PO lines, saves
* User edits PO, changes quantity on one of the lines -> the lines planned date is reset as a side effect. This will cause frustration for the user as they will easily miss the date reset, or will be painfully aware of it and complain they need to put back the date to the original value after changing a quantity on a PO line. The change saves their time and sanity. 

Please keep both in migrations of this module.

**Comment by kimirondon** [comment]
hello @FernandoRomera is this PR close to be merged?

**Comment by vincent-cowboy** [comment]
is the fix https://github.com/OCA/purchase-workflow/pull/1350 relevant to this PR?
also, there is a refactoring PR https://github.com/OCA/purchase-workflow/pull/1326 (linked to https://github.com/OCA/purchase-workflow/pull/1246)

**Comment by andreampiovesana** [comment]
merge?

**Pull request merged**